Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,158 @@
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        private const int GET_APP_ID_TIMEOUT = 2000; // milliseconds
+
+        private const string CORRELATION_ID_FORMAT = "aid-v1:{0}";
+
+        private const string APPID_QUERY_API_RELATIVE_URI_FORMAT = "api/profiles/{0}/appId";
+
+        private static ConcurrentDictionary<string, string> knownCorrelationIds = new ConcurrentDictionary<string, string>();
+
+        private static Func<string, string, Task<string>> provideAppId = FetchAppIdFromService;
+
+        /// <summary>
+        /// This is a test hook. Use this to provide your own test implementation for fetching the appId.
+        /// </summary>
+        /// <param name="appIdProviderMethod">The delegate to be called to fetch the appId</param>
+        public static void OverrideAppIdProvider(Func<string, string, Task<string>> appIdProviderMethod)
+        {
+            provideAppId = appIdProviderMethod;
+        }
+
+        /// <summary>
+        /// Retrieves the correlation id corresponding to a given instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key string.</param>
+        /// <param name="breezeEndpointAddress">Breeze endpoint to query against.</param>
+        /// <param name="correlationId">AppId corresponding to the provided instrumentation key</param>
+        /// <returns>true if correlationId was successfully retrieved; false otherwise.</returns>
+        public static bool TryGetXComponentCorrelationId(string instrumentationKey, string breezeEndpointAddress, out string correlationId)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                throw new ArgumentNullException("instrumentationKey");
+            }
+
+            var found = knownCorrelationIds.TryGetValue(instrumentationKey, out correlationId);
+
+            if (found)
+            {
+                return true;
+            }
+            else
+            {
+
+                if (breezeEndpointAddress == null)
+                {
+                    throw new ArgumentNullException("breezeEndpointAddress");
+                }
+
+                // Simplistic cleanup to guard against this becoming a memory hog.
+                if (knownCorrelationIds.Keys.Count >= MAXSIZE)
+                {
+                    knownCorrelationIds.Clear();
+                }
+
+                try
+                {
+                    // Todo: When this fails, say in the vortex endpoint case, ProfileQueryEndpont is not provided, it may perpetually keep failing.
+                    // We can possibly make it more robust by having an exponential backoff on making a call to prod endpoint. Or store failure and never query again.
+                    // Is that worth the effort?
+                    
+                    // We wait for 2 seconds to retrieve the appId. If retrieved during that time, we return success setting the correlation id.
```

Comment:
I've reduced the time out to 0 ms. At this point this should pretty much be entirely async.

I've not eliminated the timeout totally - instead have allowed the test hook to provide a non-zero timeout so that tests can work without having to make the calls twice or having to resort to pre-populating the dictionary - which may result in not hitting some correlation id formatting code.

However, the communication has to be in the user code context - because the instrumentation key required to get the app id is not known until the send telemetry call.